internal/topology/check/compatibility.go

```diff
@@ -176,10 +177,155 @@ func LocalObjectTemplateIsValid(in *clusterv1.LocalObjectTemplate, namespace str
 }
+
+// ClusterClassesAreCompatible checks the compatibility between new and old versions of a Cluster Class.
+// It checks that:
+// 1) InfrastructureCluster Templates are compatible.
+// 2) ControlPlane Templates are compatible.
+// 3) ControlPlane InfrastructureMachineTemplates are compatible.
+// 4) MachineDeploymentClasses have not been deleted and are compatible.
+func ClusterClassesAreCompatible(current, desired *clusterv1.ClusterClass) field.ErrorList {
+	var allErrs field.ErrorList
+	if current == nil {
+		return nil
+	}
+
+	// Validate InfrastructureClusterTemplate changes desired a compatible way.
+	allErrs = append(allErrs, LocalObjectTemplatesAreCompatible(current.Spec.Infrastructure, desired.Spec.Infrastructure, field.NewPath("spec", "infrastructure"))...)
+
+	// Validate control plane changes desired a compatible way.
+	allErrs = append(allErrs, LocalObjectTemplatesAreCompatible(current.Spec.ControlPlane.LocalObjectTemplate, desired.Spec.ControlPlane.LocalObjectTemplate, field.NewPath("spec", "controlplane", "localobjecttemplate"))...)
+	if desired.Spec.ControlPlane.MachineInfrastructure != nil && current.Spec.ControlPlane.MachineInfrastructure != nil {
+		allErrs = append(allErrs, LocalObjectTemplatesAreCompatible(*current.Spec.ControlPlane.MachineInfrastructure, *desired.Spec.ControlPlane.MachineInfrastructure, field.NewPath("spec", "controlplane", "machineinfrastructure"))...)
+	}
+
+	// Validate changes to MachineDeployments.
+	allErrs = append(allErrs, MachineDeploymentClassesAreCompatible(current, desired)...)
+
+	return allErrs
+}
+
+// MachineDeploymentClassesAreCompatible checks if each MachineDeploymentClass in the new ClusterClass is a compatible change from the previous ClusterClass.
+// It checks if:
+// 1) Any MachineDeploymentClass has been removed.
+// 2) If the MachineDeploymentClass.Template.Infrastructure reference has changed its Group or Kind.
+func MachineDeploymentClassesAreCompatible(current, desired *clusterv1.ClusterClass) field.ErrorList {
+	var allErrs field.ErrorList
+
+	// Ensure no MachineDeployment class was removed.
+	classes := classNamesFromWorkerClass(desired.Spec.Workers)
+	for i, oldClass := range current.Spec.Workers.MachineDeployments {
+		if !classes.Has(oldClass.Class) {
+			allErrs = append(allErrs,
+				field.Invalid(
+					field.NewPath("spec", "workers", "machineDeployments").Index(i),
+					desired.Spec.Workers.MachineDeployments,
+					fmt.Sprintf("The %q MachineDeployment class can't be removed.", oldClass.Class),
+				),
+			)
+		}
+	}
+
+	// Ensure previous MachineDeployment class was modified in a compatible way.
+	for _, class := range desired.Spec.Workers.MachineDeployments {
+		for i, oldClass := range current.Spec.Workers.MachineDeployments {
+			if class.Class == oldClass.Class {
+				// NOTE: class.Template.Metadata and class.Template.Bootstrap are allowed to change;
+
+				// class.Template.Bootstrap is ensured syntactically correct by LocalObjectTemplateIsValid.
+
```

Comment:
```suggestion
				// class.Template.Bootstrap is ensured syntactically correct by LocalObjectTemplateIsValid.
```
nit: I assume the empty lines before and after are not intended

Reply:
These were added in response to: https://github.com/kubernetes-sigs/cluster-api/pull/5644#discussion_r751203097

Reply:
Okay, then let's just keep them, not really worth discussing it in detail :)